src/joystick/windows/SDL_dinputjoystick.c

```diff
@@ -235,208 +233,102 @@ SetDIerror(const char *function, HRESULT code)
-        }
+    if (!device || !device_name) {
+        return SDL_FALSE;
     }
 
-LCleanup:
-    if (bstrNamespace) {
-        SysFreeString(bstrNamespace);
-    }
-    if (bstrDeviceID) {
-        SysFreeString(bstrDeviceID);
-    }
-    if (bstrClassName) {
-        SysFreeString(bstrClassName);
-    }
-    for (iDevice = 0; iDevice < SDL_arraysize(pDevices); iDevice++) {
-        if (pDevices[iDevice]) {
-            IWbemClassObject_Release(pDevices[iDevice]);
-        }
-    }
-    if (pEnumDevices) {
-        IEnumWbemClassObject_Release(pEnumDevices);
-    }
-    if (pIWbemLocator) {
-        IWbemLocator_Release(pIWbemLocator);
-    }
-    if (pIWbemServices) {
-        IWbemServices_Release(pIWbemServices);
+    dipstr.diph.dwSize = sizeof(dipstr);
+    dipstr.diph.dwHeaderSize = sizeof(dipstr.diph);
+    dipstr.diph.dwObj = 0;
+    dipstr.diph.dwHow = DIPH_DEVICE;
+
+    if (FAILED(IDirectInputDevice8_GetProperty(device, DIPROP_PRODUCTNAME, &dipstr.diph))) {
+        return SDL_FALSE;
     }
 
-    return bIsXinputDevice;
+    *device_name = WIN_StringToUTF8(dipstr.wsz);
+
+    return SDL_TRUE;
 }
-#endif /* 0 */
 
 static SDL_bool
-SDL_IsXInputDevice(const TCHAR *name, const GUID* pGuidProductFromDirectInput)
+QueryDevicePath(LPDIRECTINPUTDEVICE8 device, char** device_path)
 {
-    UINT i;
+    DIPROPGUIDANDPATH dippath;
 
-    if (!SDL_XINPUT_Enabled()) {
+    if (!device || !device_path) {
         return SDL_FALSE;
     }
 
-    if (SDL_tcsstr(name, TEXT(" XINPUT ")) != NULL) {
-        /* This is a duplicate interface for a controller that will show up with XInput,
-           e.g. Xbox One Elite Series 2 in Bluetooth mode.
-         */
-        return SDL_TRUE;
-    }
+    dippath.diph.dwSize = sizeof(dippath);
+    dippath.diph.dwHeaderSize = sizeof(dippath.diph);
+    dippath.diph.dwObj = 0;
+    dippath.diph.dwHow = DIPH_DEVICE;
 
-    if (SDL_memcmp(&pGuidProductFromDirectInput->Data4[2], "PIDVID", 6) == 0) {
-        Uint16 vendor_id = (Uint16)LOWORD(pGuidProductFromDirectInput->Data1);
-        Uint16 product_id = (Uint16)HIWORD(pGuidProductFromDirectInput->Data1);
-        SDL_GameControllerType type = SDL_GetJoystickGameControllerType("", vendor_id, product_id, -1, 0, 0, 0);
-        if (type == SDL_CONTROLLER_TYPE_XBOX360 ||
-            type == SDL_CONTROLLER_TYPE_XBOXONE ||
-            (vendor_id == 0x28DE && product_id == 0x11FF)) {
-            return SDL_TRUE;
-        }
+    if (FAILED(IDirectInputDevice8_GetProperty(device, DIPROP_GUIDANDPATH, &dippath.diph))) {
+        return SDL_FALSE;
     }
 
-    /* Go through RAWINPUT (WinXP and later) to find HID devices. */
-    /* Cache this if we end up using it. */
-    if (SDL_RawDevList == NULL) {
-        if ((GetRawInputDeviceList(NULL, &SDL_RawDevListCount, sizeof(RAWINPUTDEVICELIST)) == -1) || (!SDL_RawDevListCount)) {
-            return SDL_FALSE;  /* oh well. */
-        }
+    *device_path = WIN_StringToUTF8(dippath.wszPath);
```

Comment:
I guess you want `WIN_StringToUTF8W` here, because dippath.wszPath `WCHAR[]`

Reply:
fixed